ledger/streaming/client/src/source.rs

```diff
@@ -3,70 +3,92 @@
 use displaydoc::Display;
 use futures::{Stream, StreamExt};
 use mc_common::logger::{log, o, Logger};
+use mc_crypto_keys::Ed25519Public;
 use mc_ledger_streaming_api::{
+    parse_subscribe_response,
     streaming_blocks::{SubscribeRequest, SubscribeResponse},
     streaming_blocks_grpc::LedgerUpdatesClient,
-    BlockSource, Error, Result,
+    BlockStream, BlockStreamComponents, Result,
 };
-use mc_transaction_core::BlockData;
 use mc_util_grpc::ConnectionUriGrpcioChannel;
 use mc_util_uri::ConnectionUri;
-use std::{convert::TryFrom, sync::Arc, time::Duration};
+use std::{sync::Arc, time::Duration};
 
 #[derive(Display)]
 pub struct GrpcBlockSource {
     /// The gRPC client
     client: LedgerUpdatesClient,
+    /// The public key of the consensus node we're streaming from.
+    public_key: Arc<Ed25519Public>,
     /// A logger object
     logger: Logger,
 }
 
 impl GrpcBlockSource {
-    pub fn new(uri: &impl ConnectionUri, env: Arc<grpcio::Environment>, logger: Logger) -> Self {
+    pub fn new(
+        uri: &impl ConnectionUri,
+        env: Arc<grpcio::Environment>,
+        public_key: Ed25519Public,
+        logger: Logger,
+    ) -> Self {
         let logger = logger.new(o!("uri" => uri.to_string()));
         let channel =
             grpcio::ChannelBuilder::default_channel_builder(env).connect_to_uri(uri, &logger);
         let client = LedgerUpdatesClient::new(channel);
-        Self { client, logger }
+        Self {
+            client,
+            public_key: Arc::new(public_key),
```

Comment:
Do we need the Arc? afaik this key doesn't change.

Reply:
Which key is this? If this is the block signing key, then that changes when a node restarts.

Reply:
Looking at the code it appears this might be a separate key, so I am not sure where it is coming from.

Reply:
It's a new keypair that will be used to sign and verify the `BlockWithQuorumSet result` field in `SubscribeResponse`. It'll be configured per consensus node.

I put it in an `Arc` so that a reference to the key can be paired with the stream consumer, without retaining a reference to the sink, i.e. to appease the lifetime/borrow checker.

Reply:
It's intended to be the message signing key.